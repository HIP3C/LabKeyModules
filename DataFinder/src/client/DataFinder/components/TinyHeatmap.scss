
.tinyheatmap {
    height: 120px;
    position: relative;
    border: 0;
  }
  
  .tinyheatmap > svg {
    position: absolute;
    bottom: 0;
  }