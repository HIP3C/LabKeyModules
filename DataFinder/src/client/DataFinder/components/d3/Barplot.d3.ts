import * as d3 from 'd3'
import { PlotDatum } from '../../../typings/CubeData';

// ================================================================== //
/* 
This is a barplot component which takes the following arguments in 
the props: 
  * data
  * name (id for the div to stick it in)
  * width
  * height

*/
// Types

interface D3BarplotConfig {
    width: number;
    height: number;
    labels: string[];
    countMetric: string;
    barColor: string;
}


interface D3Barplot {
    create: (id: string, data: PlotDatum[], config: D3BarplotConfig) => void;
    update: (id: string, data: PlotDatum[], config: D3BarplotConfig) => void;
}


const xaxisTitle = {height: 40}

const createBarplot = (id: string, data: PlotDatum[], config: D3BarplotConfig) => {

    // Create margins
    const barplotHeight = config.height - xaxisTitle.height - 10
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
    width = config.width - margin.left - margin.right,
    height = barplotHeight - margin.top - margin.bottom;

    // Add svg for bars and separate one for x-axis to allow
    const svg = d3
        .select("#barplot-container-" + id).select("svg")
        .attr("height", barplotHeight)
        .attr("width", config.width)
        .attr("id", "svg-barplot-" + id)

    const xAxisSvg = d3.select("#xaxis-" + id).select("svg")
        .attr("width", config.width)
        .attr("height", xaxisTitle.height)


    // Get long and short versions of y-axis labels
    const labels = config.labels.map(l => {
        let short = l
        const maxLength = Math.round( margin.left / 5.5 )
        if (l.length > maxLength) short = l.slice(0,  maxLength - 3) + "..."
        return ({ label: l, shortlabel: short })
    })
    const yaxisScale = d3
        .scaleBand()
        .domain(labels.map(l => l.label))
        .range([height, 0]);

    // Add container for bars
    const barplot = svg
        .append("g")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .attr("id", "bars-" + id)
        .attr("style", "overflow:auto")

    // Add y-axis short labels
    const yaxisLabels = svg.append("g")
        .attr("id", "yaxis-labels-short-" + id)
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
    yaxisLabels.selectAll("text")
        .data(labels)
        .enter()
        .append("text")
        .attr("x", -5)
        .attr("y", (d) => {
            return yaxisScale((d.label)) + yaxisScale.bandwidth() / 1.5;
        })
        .attr("text-anchor", "end")
        .attr("font-size", ".8em")
        .text(d => d.shortlabel)

    // Insert vertical grid lines (before bars)
    const xaxisGrid = svg.insert("g", ":first-child")
        .attr("id", "xaxis-grid-" + id)
        .attr(
            "transform",
            `translate(${margin.left}, ${(height + 5 + margin.top)})`
        )

    // x-axis label container (scale changes on update)
    const xaxisLabels = xAxisSvg.append("g")
        .attr("class", "x axis")
        .attr(
            "transform",
            `translate(${margin.left}, 0)`
        )

    // x-axis title
    xAxisSvg.append("g")
        .attr(
            "transform",
            `translate(${margin.left}, 0)`
        )
        .append("text")
        .attr("x", (width) / 2)
        .attr("y", 30)
        .attr("class", "x-axis-title")
        .text(config.countMetric == "studyCount" ? "Studies" : "Participants")
        .attr("text-anchor", "middle")
        .attr("font-size", "0.8em")

    // y-axis long labels container (for hover)
    const yaxisLongLabels = svg.append("g")
        .attr("id", "yaxis-labels-" + id)
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
}

const updateBarplot = (id: string, data: PlotDatum[], config: D3BarplotConfig) => {
    // Prelims -----------------
    const svg = d3.select("#barplot-container-" + id).select("svg")
    const xAxisSvg = d3.select("#xaxis-" + id).select("svg")
    const barplot = d3.select("#bars-" + id)
    const xaxisGrid = svg.select("#xaxis-grid-" + id)
    const xaxisLabels = xAxisSvg.selectAll(".x.axis")
    // yaxisLongLabels -- a group on top of the othe
    const yaxisLongLabels = svg.select("#yaxis-labels-" + id) 

    

    // Create margins
    const barplotHeight = config.height - xaxisTitle.height - 10
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
    width = config.width - margin.left - margin.right,
    height = barplotHeight - margin.top - margin.bottom;
    
    const yaxisScale = d3
        .scaleBand()
        .domain(config.labels.map(l => l))
        .range([height, 0]);
    

    // Make the updates ------------
    // Bind data
    // Get range of data for x-axis
    const dataRange = [0, 10];
    const countMetric = config.countMetric
    data.forEach((v) => (v[countMetric] > dataRange[1]) && (dataRange[1] = v[countMetric]))
    const xaxisScale = d3
        .scaleLinear()
        .domain(dataRange).nice()
        .range([0, width]);
    const updateXaxis: any = d3.axisBottom(xaxisScale).ticks(3).tickSize(0)
    const updateXaxisGrid: any = d3.axisBottom(xaxisScale).ticks(3).tickSize(-(height + 5)).tickFormat(() => "")

    // Bind the data
    const boxes = barplot.selectAll("rect").data(data);

    const yaxisHoverLabels = yaxisLongLabels
        .selectAll(".yaxis-long-label-container")
        .data(data)

    // ---- enter ----
    // bars on the barplot
    boxes
        .enter()
        .append("rect")
        .attr("class", "rect")
        .attr("x", xaxisScale(0))
        .attr("width", function (d: PlotDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: PlotDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)
        .style("fill", config.barColor)
    
    // Labels for hover
    const labelContainers = yaxisHoverLabels
        .enter()
        .append("g")
        .attr("class", "yaxis-long-label-container") 

    labelContainers
        .append("text")
        .attr("class", "label-number")
        .attr("x", config.width - margin.left)
        .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
        .attr("text-anchor", "end")
        .text(d => d[countMetric] == null ? 0 : d[countMetric])
        .attr("font-size", ".8em")
        .attr("fill", "transparent")
        
    // Rect over the whole area around the bar to use for adding mouseover event
    labelContainers
        .append("rect")
        .attr("class", "big-rect-cover")
        .attr("x", - margin.left)
        .attr("y", (d: PlotDatum) => {return(yaxisScale(d.member))})
        .attr("width", config.width)
        .attr("height", yaxisScale.bandwidth())
        .attr("fill", "transparent")
        .attr("stroke", "transparent")
    // rect over the short label. Will become opaque to cover this label
    // and replace with the long version on mouseover. 
    labelContainers
        .append("rect")
        .attr("class", "little-rect-cover")
        .attr("x", d => d == null ? undefined : -margin.left)
        .attr("y", (d: PlotDatum) => yaxisScale(d.member))
        .attr("width", 100)
        .attr("height", yaxisScale.bandwidth())
        .attr("fill", "transparent")
        .attr("stroke", "transparent")
    // y-axis labels
    labelContainers
        .append("text")
        .attr("class", "label-text")
        .attr("x", (d: PlotDatum) => d.member.length < 18 ? -5 : - margin.left)
        .attr("y", (d: PlotDatum) => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
        .attr("text-anchor", (d: PlotDatum) => d.member.length < 18 ? "end" : "start")
        .attr("font-size", ".8em")
        .attr("fill", "transparent")
      .selectAll("tspan")
        .data((d: PlotDatum) => {
            let l = d.member
            const maxLength = Math.round( (config.
                width - margin.right - 10) / 5.5 )
            if (l.length > maxLength) {
                let splitIndex = l.indexOf(" ")
                while (splitIndex < maxLength) {
                    const next = l.indexOf(" ", splitIndex + 1)
                    if (next > -1 && next < maxLength) { splitIndex = next } else { break }
                }
                let newlabel = [l.slice(0, splitIndex), l.slice(splitIndex + 1)]
                if (newlabel[1].length > maxLength) newlabel[1] = newlabel[1].slice(0, maxLength - 4) + "..."
                return (newlabel)
            } else { return ([l]) }
        })
        .enter()
        .append("tspan")
        .attr("x", (d, i) => d.length < 18 && i === 0 ? -5 : - margin.left)
        .attr("dy", (d, i, data) => i > 0 ? "10" : data.length > 1 ? "-5" : "0")
        .text(d => d)
    
    

    // ---- update ----
    boxes
        .transition()
        .duration(500)
        .attr("width", function (d: PlotDatum) {
            return xaxisScale(d[countMetric]);
        })

    yaxisHoverLabels
        .selectAll(".label-number")
        .data(d => [d])
        .text(d => d[countMetric] == null ? 0 : d[countMetric])

    xaxisLabels.transition()
        .duration(500)
        .call(updateXaxis)
    xaxisGrid.transition()
        .duration(500)
        .call(updateXaxisGrid)
        .selectAll("line").attr("stroke", "#ddd")
    d3.selectAll(".domain").remove() // Removes the line on the bottom of the plot

    // ---- exit ----
    boxes.exit().remove();

    // ----- Mouseover event -----
    yaxisLongLabels.selectAll("g")
        .on("mouseover", function (d) {
            d3.select(this)
                .selectAll(".label-text")
                .transition()
                .duration(100)
                .attr("fill", "black")
                .attr("font-weight", "bold")
            d3.select(this)
                .selectAll(".label-number")
                .transition()
                .duration(100)
                .attr("fill", "black")
                .attr("font-weight", "bold")
            d3.select(this)
                .selectAll(".little-rect-cover")
                .transition()
                .duration(100)
                .attr("fill", "white")
        })
        .on("mouseout", function (d) {
            d3.select(this)
                .selectAll(".label-text")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
                .attr("font-weight", "normal")
            d3.select(this)
                .selectAll(".label-number")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
                .attr("font-weight", "normal")
            d3.select(this)
                .selectAll(".little-rect-cover")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
        })

}

export const D3Barplot: D3Barplot = {
    create: createBarplot,
    update: updateBarplot
}
