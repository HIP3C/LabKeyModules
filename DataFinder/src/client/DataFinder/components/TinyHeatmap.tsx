import * as React from 'react';
import { drawTinyHeatmap } from './d3/TinyHeatmap.d3'
import { TinyHeatmapInfo } from '../../typings/StudyCard'
// Barplot ---------------------------------------- //
export interface TinyHeatmapProps {
    name: string;
    heatmapInfo: TinyHeatmapInfo
}

export const TinyHeatmap: React.FC<TinyHeatmapProps> = (props) => {

    React.useEffect(() => {
        if (props.heatmapInfo.data.length > 0 && props.heatmapInfo.assays.length > 0) {
            drawTinyHeatmap(props);
        }
    }, []);

    return (
        <div className={"tinyheatmap"} >
            <svg id={"tinyheatmap-" + props.name}></svg>
        </div>
    );


}
