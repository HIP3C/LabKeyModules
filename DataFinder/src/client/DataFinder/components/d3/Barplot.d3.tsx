import * as d3 from 'd3'
import { CubeDatum } from '../../../typings/CubeData';
import { entries } from 'd3';

// ================================================================== //
/* 
This is a barplot component which takes the following arguments in 
the props: 
  * data
  * name (id for the div to stick it in)
  * width
  * height

*/
// Types
interface DrawBarplotProps {
    data: CubeDatum[];
    name: string;
    width: number;
    height: number;
    labels: string[];
    countMetric: string;
    barColor: string;
}


export function drawBarplot(props: DrawBarplotProps) {

    const data = props.data;
    const name = props.name;

    const dataRange = [0, 10];
    const countMetric = props.countMetric
    data.forEach((v) => (v[countMetric] > dataRange[1]) && (dataRange[1] = v[countMetric]))

    const totalHeight = Math.max(195, 15 * props.labels.length + 20)

    const labels = props.labels.map(l => {
        let short = l
        if (l.length > 18) short = l.slice(0, 14) + "..."
        return ({ label: l, shortlabel: short })
    })
    // debuggers
    // const newLabels = [];

    // data.map((e, i) => {
    //     newLabels.push(e.label);
    //     // if (e.value > dataRange[1]) dataRange[1] = e.value;
    // });

    const svg = d3
        .select("#barplot-container-" + name).select("svg")
        .attr("height", totalHeight)
        .attr("width", 220)
        .attr("id", "barplot-container-" + name)

    const xAxisSvg = d3.select("#xaxis-" + name).select("svg")
        .attr("width", props.width)
        .attr("height", 40)


    // Create margins
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
        width = props.width - margin.left - margin.right,
        height = totalHeight - margin.top - margin.bottom;

    // Set scales using options

    const xaxisScale = d3
        .scaleLinear()
        .domain(dataRange).nice()
        .range([0, width]);

    const yaxisScale = d3
        .scaleBand()
        .domain(labels.map(l => l.label))
        .range([height, 0]);

    // Create body and axes
    // svg.append("g")
    //     .call(d3.axisLeft(yaxisScale));

    // svg.append("g")
    //     .call(d3.axisBottom(xaxisScale));

    let barplot: d3.Selection<SVGElement, any, HTMLElement, any>;
    let xaxisGrid, xaxisLabels, yaxisLabels
    if (!svg.selectAll("g").empty()) {
        svg.select("#xaxis-labels-" + name).remove()
        xaxisGrid = svg.select("#xaxis-grid-" + name)
        xaxisLabels = xAxisSvg.selectAll(".x.axis")
        barplot = svg.select("#barplot" + name)
    } else {
        barplot = svg
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .attr("id", "barplot" + name)
            .attr("style", "overflow:auto")

        yaxisLabels = svg.append("g")
            .attr("id", "yaxis-labels-short-" + name)
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
        yaxisLabels.selectAll("text")
            .data(labels)
            .enter()
            .append("text")
            .attr("x", -5)
            .attr("y", (d) => {
                return yaxisScale((d.label)) + yaxisScale.bandwidth() / 1.5;
            })
            .attr("text-anchor", "end")
            .attr("font-size", ".8em")
            .text(d => d.shortlabel)

        const yaxisLongLabels = svg.append("g")
            .attr("id", "yaxis-labels-" + name)
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        const labelContainers = yaxisLongLabels
            .selectAll("g")
            .data(data)
            .enter()
            .append("g")

        labelContainers
            .selectAll(".big-rect-cover")
            .data(d => [d])
            .enter()
            .append("rect")
            .attr("class", "big-rect-cover")
            .attr("x", - margin.left)
            .attr("y", d => yaxisScale(d.member))
            .attr("width", props.width)
            .attr("height", yaxisScale.bandwidth())
            .attr("fill", "transparent")
            .attr("stroke", "transparent")
        labelContainers
            .selectAll(".little-rect-cover")
            .data(d => [d])
            .enter()
            .append("rect")
            .attr("class", "little-rect-cover")
            .attr("x", d => d == null ? undefined : -margin.left)
            .attr("y", d => yaxisScale(d.member))
            .attr("width", 100)
            .attr("height", yaxisScale.bandwidth())
            .attr("fill", "transparent")
            .attr("stroke", "transparent")
        labelContainers.selectAll(".label-text")
            .data(d => [d])
            .enter()
            .append("text")
            .attr("class", "label-text")
            .attr("x", d => d.member.length < 18 ? -5 : - margin.left)
            .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
            .attr("text-anchor", d => d.member.length < 18 ? "end" : "start")
            .attr("font-size", ".8em")
            .attr("fill", "transparent")
            .text(d => d.member)
        labelContainers.selectAll(".label-number")
            .data(d => [d])
            .enter()
            .append("text")
            .attr("class", "label-number")
            .attr("x", props.width - margin.left - 30)
            .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
            .attr("text-anchor", "end")
            .text(d => d[countMetric] == null ? 0 : d[countMetric])
            .attr("font-size", ".8em")
            .attr("fill", "transparent")

        yaxisLongLabels.selectAll("g")
            .on("mouseover", function (d) {
                d3.select(this)
                    .selectAll(".label-text")
                    .transition()
                    .duration(100)
                    .attr("fill", "black")
                    .attr("font-weight", "bold")
                d3.select(this)
                    .selectAll(".label-number")
                    .transition()
                    .duration(100)
                    .attr("fill", "black")
                    .attr("font-weight", "bold")
                d3.select(this)
                    .selectAll(".little-rect-cover")
                    .transition()
                    .duration(100)
                    .attr("fill", "white")
            })
            .on("mouseout", function (d) {
                d3.select(this)
                    .selectAll(".label-text")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
                    .attr("font-weight", "normal")
                d3.select(this)
                    .selectAll(".label-number")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
                    .attr("font-weight", "normal")
                d3.select(this)
                    .selectAll(".little-rect-cover")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
            })

        xaxisGrid = svg.insert("g", ":first-child")
            .attr("id", "xaxis-grid-" + name)
            .attr(
                "transform",
                `translate(${margin.left}, ${(height + 5 + margin.top)})`
            )

        xaxisLabels = xAxisSvg.append("g")
            .attr("class", "x axis")
            .attr(
                "transform",
                `translate(${margin.left}, 0)`
            )

        xAxisSvg.append("g")
            .attr(
                "transform",
                `translate(${margin.left}, 0)`
            )
            .append("text")
            .attr("x", (width) / 2 )
            .attr("y", 30)
            .text(props.countMetric == "studyCount" ? "Studies" : "Participants")
            .attr("text-anchor", "middle")
            .attr("font-size", "0.8em")
    }

    xaxisGrid.transition()
        .duration(500)
        .call(d3.axisBottom(xaxisScale).ticks(3).tickSize(-(height + 5)).tickFormat(() => ""))
        .selectAll("line").attr("stroke", "#ddd")
    xaxisLabels.transition()
        .duration(500)
        .call(d3.axisBottom(xaxisScale).ticks(3).tickSize(0))
    // svg.append("g")
    //     .attr("id", "xaxis-labels-" + name)
    //     .attr(
    //         "transform",
    //         `translate(${margin.left}, ${(height + margin.top)})`
    //     )
    //     .call(d3.axisBottom(xaxisScale).tickValues([Math.round((dataRange[1] + 5) / 10) * 10]).tickSize(-height))
    //     .selectAll("line").attr("stroke", "#bcbcbc").attr("stroke-width", "3")

    d3.selectAll(".domain").remove()




    // add data
    const boxes = barplot.selectAll("rect").data(data);
    boxes
        .enter()
        .append("rect")
        .attr("class", "rect")
        .attr("x", xaxisScale(0))
        .attr("width", function (d: CubeDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: CubeDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)
        .style("fill", props.barColor)
    boxes
        .transition()
        .duration(500)
        .attr("x", xaxisScale(0))
        .attr("width", function (d: CubeDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: CubeDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)

    boxes.exit().remove();
}

