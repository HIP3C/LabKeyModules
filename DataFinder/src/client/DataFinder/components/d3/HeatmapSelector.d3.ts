import * as d3 from 'd3'
import { HeatmapDatum, IAssayData, PlotDatum, Filter } from '../../../typings/CubeData';
import { HeatmapProps } from '../HeatmapSelector';
import { AxisDatum } from '../HeatmapSelector';
import { Axis, select } from 'd3';
import { List } from 'immutable';
// ================================================================== //
/* 
This is a heatmap component which was translated from an r2d3 fuction.
It takes the following arguments:
  * data (array of objects, each one representing a grid rectangle.
          objects should have the following format:
         {"assay": "",
          "id": "10b", // a unique id for this square
          "participantCount": 10,:
          "participantList": [], // array of participant IDs
          "studyCount": 2,
          "studyList": [], // array of study IDs
          "timepoint": "0"
          }
  )
  * name (class name for the div to stick it in)
  * width (width for the plot)
  * height (height for the plot)
  * options (object with the following format:
      {"breaks": [], //(numerical array with breaks for coloring)
       "colors": [], //(array of hex colors, where length = breaks.length + 2)
       "xaxis": [], //(character array with labels for the x-axis)
       "yaxis": [], //(character array with labels for the y-axis)
       "selected": [] //(elements that are currently selected)
      }
     )
*/



export const drawHeatmapSelector = (props: HeatmapProps) => {
  // if (props.showSampleType) debugger
  const data = props.data;
  const name = props.name;
  const selected = props.selected;

  // if (data.length > 0) debugger

  d3.select("." + name + " > *").remove();
  var svg = d3
    .select("." + name)
    .append("svg")
    .attr("height", props.height)
    .attr("width", props.width)
    .attr("id", "heatmap-" + name);

  // debugger
  // x, y, width, top, right, left, height, bottom

  // Create margins
  var margin = { top: 15, right: 15, bottom: 40, left: 135 },
    width = props.width - margin.left - margin.right,
    height = props.height - margin.top - margin.bottom;

  // Define the div for the tooltip
  var tooltip = d3.select("#heatmap-label");
  var arrow = d3.select(".arrow-down");

  // Set scales using options
  // color scale
  var colorScale = d3
    .scaleThreshold<number, string>()
    .domain(props.breaks)
    .range(props.colors);

  var xaxisScale = d3
    .scaleBand()
    .domain(props.xaxis.map((e) => e.label))
    .range([0, width]);

  var yaxisScale = d3
    .scaleBand()
    .domain(props.yaxis.map((e) => e.label))
    .range([0, height]);

  // Create body and axes
  let heatmap: d3.Selection<SVGElement, any, HTMLElement, any>;

  if (svg.selectAll("g").empty()) {
    heatmap = svg
      .append("g")
      .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
      .attr("id", "heatmap-" + name);
    heatmap.append("rect")
      .attr("id", "heatmap-bounding-rect")
      .attr("x", 0)
      .attr("y", 0)
      .attr("width", width)
      .attr("height", height)
      .attr("fill", "white")
    var xaxislabels = svg
      .append("g")
      .attr("id", "xaxis-labels")
      .attr(
        "transform",
        "translate(" + margin.left + ", " + (height + margin.top) + ")"
      );
    //.call(d3.axisBottom(xaxisScale));

    var yaxislabels = svg
      .append("g")
      .attr("id", "yaxis-labels")
      .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
    // .call(d3.axisLeft(yaxisScale));
  } else {
    heatmap = svg.select("#heatmap"),
      xaxislabels = svg.select("#xaxis-labels"),
      yaxislabels = svg.select("#yaxis-labels");
  }

  

  var yaxistext = yaxislabels.selectAll("text").data(props.yaxis);
  yaxistext
    .enter()
    .append("text")
    .attr("x", (d) => {
      return xaxisScale("<0") - 5;
    })
    .attr("y", (d) => {
      return yaxisScale((d.label)) + yaxisScale.bandwidth() / 2;
    })
    .attr("text-anchor", "end")
    .attr("font-size", ".8em")
    .attr("dominant-baseline", "central")
    .text((d) => {
      return d.label;
    });

  

  

  var xaxis = xaxislabels.selectAll("labels").data(props.xaxis);

  if (svg.select("#xaxis-title").empty()) {
    xaxis
      .enter()
      .append("text")
      .attr("x", (d) => {
        return xaxisScale(d.label) + xaxisScale.bandwidth() / 2;
      })
      .attr("y", (d) => {
        return 17;
      })
      .attr("text-anchor", "middle")
      .attr("font-size", ".8em")
      .text((d) => {
        return d.label;
      });

    

    xaxislabels
      .append("text")
      .attr("id", "xaxis-title")
      .attr("x", width / 2)
      .attr("y", "35")
      .attr("text-anchor", "middle")
      .text("Study Day");
  }

  // add data
  var boxes = heatmap.selectAll(".df-heatmap-box-"+name).data(data);
  boxes
    .enter()
    .append("rect")
    .attr("class", "df-heatmap-box-"+name)
    .attr("x", (d) => {
      return xaxisScale(d.x) + 3;
    })
    .attr("width", xaxisScale.bandwidth() - 3)
    .attr("y", (d) => {
      return yaxisScale(d.y) + 1;
    })
    .attr("height", yaxisScale.bandwidth() - 3)
    .style("fill", (d) => {
      return colorScale(d.participantCount);
    })
    .attr("stroke-width", "1px")
    .attr("stroke", function (d: HeatmapDatum<Filter>) {
        let s = false
        if (props.showSampleType && selected.getIn(["Assay", "SampleType"])) {
          if (selected.getIn(["Assay", "SampleType", "members"]).includes(d.data.member)) s = true
        } else if (selected.getIn(["Assay", "Timepoint"])) {
          if (selected.getIn(["Assay", "Timepoint", "members"]).includes(d.data.member)) s = true 
        }
        if (s) return "black"
      return ("#e5e5e5")
    })
    .on("mouseover", function (d, i) {
      // get position
      var coord = this.getBoundingClientRect();
      // Tooltip coordinates
      var r = (coord.right + window.scrollX);
      var t = coord.top + window.scrollY;
      // Change style
      d3.select(this)
        .attr("stroke-width", "3px")
        .attr("z-index", "10000");
      // Tooltip
      tooltip
        .transition()
        .duration(100)
        .style("opacity", 0.9);
      arrow
        .transition()
        .duration(100)
        .style("opacity", 0.9);
      arrow
        .style("left", r - (xaxisScale.bandwidth() / 2 - 5) + "px")
        .style("top", t - 12 + "px");
      tooltip
        .html(
          d.participantCount +
          " participants <br>" +
          d.studyCount +
          " studies <br>" +
          d.y +
          " at day " +
          d.x
        )
        .style("left", r - xaxisScale.bandwidth() / 2 + "px")
        .style("top", t - 50 - 12 + "px");
    })
    .on("mouseout", function (d, i) {
      // Reset to original
      d3.select(this)
        .attr("stroke-width", "1px")
        .attr("z-index", 10);
      tooltip
        .transition()
        .duration(100)
        .style("opacity", 0);
      arrow
        .transition()
        .duration(100)
        .style("opacity", 0);
    })
    .on("click", function (d, i) {
      props.handleClick(d.data);
    });

  boxes
    .transition()
    .duration(300)
    .attr("x", (d) => {
      return xaxisScale(d.x);
    })
    .attr("width", xaxisScale.bandwidth() - 1)
    .attr("y", (d) => {
      return yaxisScale(d.y);
    })
    .attr("height", yaxisScale.bandwidth() - 1)
    .style("fill", (d) => {
      return colorScale(d.participantCount);
    })
    .attr("stroke", (d) => {
      return "transparent";
      // }
    });

  boxes.exit().remove();
  yaxistext.exit().remove();
}