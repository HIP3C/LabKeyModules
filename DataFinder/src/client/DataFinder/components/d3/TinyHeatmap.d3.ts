import * as d3 from 'd3'
import { TinyHeatmapProps } from '../TinyHeatmap';
import { HeatmapDatum } from '../../../typings/CubeData'

// ================================================================== //

// Types



const drawTinyHeatmap = (props: TinyHeatmapProps) => {
    const heatmapInfo = props.heatmapInfo;
    const name = props.name;
    

    const svg = d3
        .select("#tinyheatmap-" + name)
        .attr("height", heatmapInfo.height)
        .attr("width", heatmapInfo.width)
        .attr("id", "tinyheatmap-" + name);

    // Create margins
    const margin = { top: 5, right: 0, bottom: 30, left: 50 },
        width = heatmapInfo.width - margin.left - margin.right,
        height = heatmapInfo.height - margin.top - margin.bottom;

    // Set scales using options
    const yaxisScale = heatmapInfo.yaxisScale
    const xaxisScale = heatmapInfo.xaxisScale
    const colorScale = heatmapInfo.colorScale


    // Create body and axes

    const heatmap = svg
        .append("g")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .attr("id", "tinyheatmap" + name);
    const xaxislabels = svg
        .append("g")
        .attr("id", "xaxis-labels-" + name)
        .attr(
            "transform",
            "translate(" + margin.left + ", " + (height + margin.top) + ")"
        );
    // .call(d3.axisBottom(xaxisScale));

    const yaxislabels = svg
        .append("g")
        .attr("id", "yaxis-labels-" + name)
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .call(d3.axisLeft(yaxisScale));

    xaxislabels.call(d3.axisBottom(xaxisScale))
        .selectAll("text")
        .attr("y", 5)
        .attr("x", -5)
        .attr("transform", "rotate(-45)")
        .style("text-anchor", "end")
        .attr("font-size", ".9em");

    // add data
    const boxes = heatmap.selectAll("rect").data(heatmapInfo.data);
    boxes
        .enter()
        .append("rect")
        .attr("x", (d: HeatmapDatum<any>) => {
            return xaxisScale(d.x);
        })
        .attr("width", xaxisScale.bandwidth() - 1)
        .attr("y", function (d: HeatmapDatum<any>) {
            return yaxisScale(d.y);
        })
        .attr("height", yaxisScale.bandwidth() - 1)
        .style("fill", function (d: HeatmapDatum<any>) {
            if (d.participantCount === 0 || d.y === undefined) return "transparent";
            return colorScale(d.participantCount);
        })
        .attr("stroke-width", "1px")
        .attr("stroke", "transparent");
}

export const D3TinyHeatmap = {drawTinyHeatmap}
