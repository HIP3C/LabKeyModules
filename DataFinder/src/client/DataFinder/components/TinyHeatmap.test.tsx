import React from 'react'
import { createTinyHeatmapConsts , TinyHeatmap, createTinyHeatmapYaxisScale} from './TinyHeatmap'
import * as renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import { D3TinyHeatmap } from './d3/TinyHeatmap.d3'


describe("createTinyHeatmapConsts", () => {
    const thc = createTinyHeatmapConsts();

    test("width", () => {
        expect(thc).toHaveProperty("width", 260)
    })

    test("colorScale", () => {
        expect(thc).toHaveProperty("colorScale")
        expect(thc.colorScale(11)).toBe("#A1D99B")
    })

    test("xaxisScale", () => {
        expect(thc).toHaveProperty("xaxisScale")
        expect(thc.xaxisScale("<0")).toBeCloseTo(0)
    })

})

describe("createTinyHeatmapYaxisScale", () => {
    const scale = createTinyHeatmapYaxisScale(["assay1", "assay2"])

    test("scale", () => {
        expect(scale.bandwidth()).toBeCloseTo(10)
        expect(scale.range()).toEqual([0, 20])
        expect(scale("assay1")).toBeCloseTo(0)
    })

})

describe("<TinyHeatmap />", () => {
    const consts = createTinyHeatmapConsts()
    const props = {
        name: "test", 
        heatmapInfo: {
            data: [
                {   x: "0",
                    y: "HAI",
                    participantCount: 15,
                    studyCount: 1,
                    data: undefined},
                {
                    x: "1",
                    y: "GE",
                    participantCount: 30,
                    studyCount: 1,
                    data: undefined
                }
            ],
            assays: ["HAI", "GE"],
            height: 50,
            width: consts.width,
            yaxisScale: createTinyHeatmapYaxisScale(["HAI", "GE"]),
            xaxisScale: consts.xaxisScale,
            colorScale: consts.colorScale
        }
    }

    test("Renders", () => {
        const tree = renderer.create(
            <TinyHeatmap {...props} />
        )
        expect(tree).toMatchSnapshot()
    })
    test("performs plotting side effect", () => {
        const useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
        const drawTinyHeatmap = jest.spyOn(D3TinyHeatmap, "drawTinyHeatmap")

        const heatmap = shallow(<TinyHeatmap {...props}/>)
        expect(drawTinyHeatmap).toHaveBeenCalled()
        expect(drawTinyHeatmap).toHaveBeenCalledTimes(1)

        drawTinyHeatmap.mockClear()
    })
    test("doesn't plot when there's no data", () => {
        const useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
        const drawTinyHeatmap = jest.spyOn(D3TinyHeatmap, "drawTinyHeatmap")

        props.heatmapInfo.data = []
        const heatmap = shallow(<TinyHeatmap {...props}/>)
        expect(drawTinyHeatmap).not.toHaveBeenCalled()
    })
})