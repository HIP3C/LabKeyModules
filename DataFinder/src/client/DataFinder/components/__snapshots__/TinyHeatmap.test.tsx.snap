// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TinyHeatmap /> Renders 1`] = `
<div
  className="tinyheatmap"
>
  <svg
    id="tinyheatmap-test"
  />
</div>
`;
